src/honeysql_postgres/format.cljc

```diff
@@ -143,11 +145,40 @@
            util/get-first
            sqlf/to-sql)))
 
-(defmethod format-clause :with-columns [[_ columns] _]
-  (sqlf/paren-wrap (->> columns
-                        util/get-first
-                        (map #(sqlf/space-join (map sqlf/to-sql %)))
-                        sqlf/comma-join)))
+(def ^:private constraints-format-map
+  {:primary-key "PRIMARY KEY"
+   :unique      "UNIQUE"})
+
+(defn- format-constraint-clause
+  [constraint]
+  (let [[constraint-type constraint-args] constraint]
+    (when (contains? constraints-format-map constraint-type)
+      (str (get constraints-format-map constraint-type)
+           (sqlf/paren-wrap (->> constraint-args
+                                 (map sqlf/to-sql)
+                                 sqlf/comma-join))))))
+
+(defn- format-columns-clause
+  [columns]
+  (->> columns
+       util/get-first
+       (map #(sqlf/space-join (map sqlf/to-sql %)))
+       sqlf/comma-join))
+
+(defmethod format-clause :with-columns [[_ columns] complete-sql-map]
+  (when-not (seq (:constraints complete-sql-map))
+    (sqlf/paren-wrap (format-columns-clause columns))))
+
+(defmethod format-clause :constraints [[_ constraints-seq] complete-sql-map]
+  (let [columns (:with-columns complete-sql-map)]
+    (when (and (seq columns)
```

Comment:
Filtering constraints before formatting the columns and constraints would form the basis for adding a trailing comma after the columns and between constraints. If the validation of the constraint type is done after adding the trailing comma, and an unsupported constraint is found, the formatted query produced is invalid because of the trailing comma. Eg:

```
(is (= ["CREATE TABLE products  (product_no integer, name text, product_sku integer, )"]
       (-> (create-table :products)
           (with-columns [[:product_no :integer]
                          [:name :text]
                          [:product_sku :integer]])
           (constraints [[:foreign-key [:product_no]]])
           sql/format)))
```
This also happens if there is more than one constraint for the same reason as this:

```
(clojure.string/join ", " ["a" nil "b"])
;; => "a, , b"
```

The nil here is added in when [this](https://github.com/nilenso/honeysql-postgres/pull/60/files#diff-f81b5b6904b08f9c50a240ecab26f660e4f13c113064b4519324a08f69ac62f8R155) isn't satisfied.

I propose filtering out all the invalid / unsupported constraints before formatting them. 

```
(defmethod format-clause :constraints [[_ [constraints]] complete-sql-map]
  (let [[columns]   (:with-columns complete-sql-map)
        constraints (filter (fn [[constraint-type constraint-args]]
                              (contains? constraints-format-map constraint-type)) constraints)]
    (sqlf/paren-wrap
     (str (when (seq columns) (format-columns-clause columns))
          (when (seq constraints) (str ", " 
                                                        (sqlf/comma-join (map format-constraint-clause
                                                                                              constraints))))))))
```

Reply:
Ahh! Thanks for that. I should have paid attention to trailing commas. I have done what you have suggested in the code block above and it works.

I've added a 4th test case as well which ignores invalid/unsupported constraints.